codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoStackStepMiddlewareGenerator.java

```diff
@@ -0,0 +1,329 @@
+            .build();
+    private static final SymbolReference METADATA_TYPE = SymbolReference.builder()
+            .symbol(SymbolUtils.createValueSymbolBuilder("middleware.Metadata")
+                    .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                    .build())
+            .build();
+
+    private final Symbol middlewareSym;
+    private final String handleMethodName;
+    private final SymbolReference inputType;
+    private final SymbolReference outputType;
+    private final SymbolReference handlerType;
+
+    /**
+     * Creates a new middleware generator with the given builder definition
+     *
+     * @param builder the builder to create the generator with
+     */
+    public GoStackStepMiddlewareGenerator(Builder builder) {
+        this.middlewareSym = SymbolUtils.createPointableSymbolBuilder(builder.identifier).build();
+        this.handleMethodName = builder.handleMethodName;
+        this.inputType = builder.inputType;
+        this.outputType = builder.outputType;
+        this.handlerType = builder.handlerType;
+    }
+
+    /**
+     * Create a new SerializeStep middleware generator with the provided type name
+     *
+     * @param identifier is the type name to identify the middleware
+     * @return the middleware generator
+     */
+    public static GoStackStepMiddlewareGenerator createSerializeStepMiddleware(String identifier) {
+        return createMiddleware(identifier,
+                "HandleSerialize",
+                createSmithyMiddlewarePackageReference("middleware.SerializeInput"),
+                createSmithyMiddlewarePackageReference("middleware.SerializeOutput"),
+                createSmithyMiddlewarePackageReference("middleware.SerializeHandler"));
+    }
+
+    /**
+     * Create a new DeserializeStep middleware generator with the provided type name
+     *
+     * @param identifier is the type name to identify the middleware
+     * @return the middleware generator
+     */
+    public static GoStackStepMiddlewareGenerator createDeserializeStepMiddleware(String identifier) {
+        return createMiddleware(identifier,
+                "HandleDeserialize",
+                createSmithyMiddlewarePackageReference("middleware.DeserializeInput"),
+                createSmithyMiddlewarePackageReference("middleware.DeserializeOutput"),
+                createSmithyMiddlewarePackageReference("middleware.DeserializeHandler"));
+    }
+
+    /**
+     * Generates a new step middleware generator
+     *
+     * @param identifier        the name of the middleware type
+     * @param handlerMethodName identifier method name to be implemented
+     * @param inputType         the middleware input type
+     * @param outputType        the middleware output type
+     * @param handlerType       the next handler type
+     * @return the middleware generator
+     */
+    public static GoStackStepMiddlewareGenerator createMiddleware(String identifier,
```

Comment:
Hate to be pedantic, but the first should be on its own line too, indented only one over from the function declaration itself.

Reply:
As an example, here's how we'd likely write this function signature:

```java
public static GoStackStepMiddlewareGenerator createMiddleware(
        String identifier,
        String handlerMethodName,
        SymbolReference inputType,
        SymbolReference outputType,
        SymbolReference handlerType
) {
```

This helps from having hard to grok really long lines of WS if you have a long main signature that takes a long argument.